### PR TITLE
Factor out debugPaintPadding and test it

### PR DESCRIPTION
**Pull request by Hixie**
I plan to use this to implement similar logic in SliverPadding.

To make this easier to test I extended the paints matcher to accept a
function that takes a canvas. While I was at it I also made it accept
a Finder, it'll go and find the render object for you.

Also added support for paints..path and fixed some grammar in the
error messages.

Also improved the docs for debugPaint*.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/debug.dart [hunk 1]
```diff
@@ -105,6 +115,34 @@ List<String> debugDescribeTransform(Matrix4 transform) {
   return matrix;
 }
 
+void _debugDrawDoubleRect(Canvas canvas, Rect outerRect, Rect innerRect, Color color) {
+  final Path path = new Path()
+    ..addRect(outerRect)
+    ..addRect(innerRect);
+  final Paint paint = new Paint()
+    ..color = color;
+  canvas.drawPath(path, paint);
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/debug.dart]: It's probably more efficient to issue two `drawRect` commands.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/rendering/debug.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/debug.dart]: The idea here is to punch a hole in the middle, not draw two overlapping rectangles. I may have gotten the winding rule we're using wrong, though, in which case it work work. I'll check that this actually works before landing.

**Comment by abarth** [comment]
Ah, that makes sense.  Like DRRect but without the roundness.

**Comment by Hixie** [comment]
Right.

**Pull request merged**